compiler/rustc_parse/src/parser/item.rs

```diff
@@ -789,16 +789,49 @@ impl<'a> Parser<'a> {
         }
     }
 
-    /// Is this an `(unsafe auto? | auto) trait` item?
-    fn check_auto_or_unsafe_trait_item(&mut self) -> bool {
-        // auto trait
-        self.check_keyword(kw::Auto) && self.is_keyword_ahead(1, &[kw::Trait])
-            // unsafe auto trait
-            || self.check_keyword(kw::Unsafe) && self.is_keyword_ahead(1, &[kw::Trait, kw::Auto])
+    /// Is this item an unsafe trait, an auto trait, have an impl restriction,
+    /// or any combination thereof?
+    fn is_trait_item_with_modifiers(&mut self) -> bool {
```

Comment:
This function is quite a weird one... not sure how to improve it but ehhh.

Also the documentation doesn't sound right to me, specifically the "have an impl restriction". Maybe ``Is this an `[ impl(in path)? ]? unsafe? auto? trait` item?``.

Also note that the previous version returned true on `unsafe auto eof` (which was not documented but is true)